crates/flox-rust-sdk/src/models/environment.rs

```diff
@@ -1,83 +1,928 @@
 // don't forget to update the man page
 pub const DEFAULT_MAX_AGE_DAYS: u32 = 90;
 
-pub enum CommonEnvironment<'flox, Git: GitProvider> {
-    Named(floxmeta::environment::Environment<'flox, Git, ReadOnly<Git>>),
-    Project(project::environment::Environment<'flox, Git, ReadOnly<Git>>),
+pub enum InstalledPackage {
+    Catalog(FloxTriple, CatalogEntry),
+    FlakeAttribute(FlakeAttribute, CatalogEntry),
+    StorePath(StorePath),
 }
 
-impl<'flox, Git: GitProvider> CommonEnvironment<'flox, Git> {
-    /// get a flake attribute for the environment
-    /// todo flake_attribute should be constructed earlier
-    pub async fn flake_attribute(
+#[async_trait]
+pub trait Environment {
+    /// Build the environment and create a result link as gc-root
+    async fn build(
         &self,
-    ) -> Result<FlakeAttribute, EnvironmentError<GenerationError<Git>, ParseInstallableError>> {
-        match self {
-            CommonEnvironment::Named(n) => n
-                .flake_attribute(Default::default())
-                .await
-                .map_err(EnvironmentError::Named),
-            CommonEnvironment::Project(p) => p.flake_attribute().map_err(EnvironmentError::Project),
-        }
+        nix: &NixCommandLine,
+        system: impl AsRef<str> + Send,
+    ) -> Result<(), EnvironmentError2>;
+
+    /// Install packages to the environment atomically
+    async fn install(
+        &mut self,
+        packages: impl IntoIterator<Item = FloxPackage> + Send,
+        nix: &NixCommandLine,
+        system: impl AsRef<str> + Send,
+    ) -> Result<(), EnvironmentError2>;
+
+    /// Uninstall packages from the environment atomically
+    async fn uninstall(
+        &mut self,
+        packages: impl IntoIterator<Item = FloxPackage> + Send,
+        nix: &NixCommandLine,
+        system: impl AsRef<str> + Send,
+    ) -> Result<(), EnvironmentError2>;
+
+    /// Activate this environment
+    async fn activate(&self, nix: &NixCommandLine) -> Result<(), EnvironmentError2>;
+
+    /// Atomically edit this environment, ensuring that it still builds
+    async fn edit(&self) -> Result<(), EnvironmentError2>;
+
+    async fn catalog(
+        &self,
+        nix: &NixCommandLine,
+        system: impl AsRef<str> + Send,
+    ) -> Result<EnvCatalog, EnvironmentError2>;
+
+    /// Return the [EnvironmentRef] for the environment for identification
+    fn environment_ref(&self) -> &EnvironmentRef;
+
+    /// Returns the environment owner
+    fn owner(&self) -> Option<EnvironmentOwner>;
+
+    /// Returns the environment name
+    fn name(&self) -> EnvironmentName;
+
+    /// Delete the Environment
+    fn delete(self) -> Result<(), EnvironmentError2>;
+}
+
+#[async_trait]
+pub trait TemporaryEnvironment {
+    async fn set_environment(
+        &mut self,
+        mut flox_nix_content: impl std::io::Read + Send,
+        nix: &NixCommandLine,
+        system: impl AsRef<str> + Send,
+    ) -> Result<(), EnvironmentError2>;
+}
```

Comment:
Does this really need to be a trait? I thought we want to keep the whole temporary/transaction part out of the interface.